asg-passes/src/dotify/expression.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+use super::*;
+
+use leo_asg::*;
+
+use petgraph::graph::NodeIndex;
+
+type M = Fixed<NodeIndex>;
+
```

Comment:
I agree, there's a lot of repetitive code. 
It can definitely be simplified; I thought that it might be best to keep everything explicit in case someone would like to tweak how they would like the ASG to be rendered. In its current implementation, if someone wants to make a change to a particular type of ASG node, they just have to look at modify code in the corresponding function.
What do you think? Is that a good enough justification for verbosity?

Reply:
In the cases where custom tweaking is desired, they can just inline the helper functions in the specific instance and it won't take too much time, but I don't think there will be too much ad-hoc customization over time. The "but new requirements will make us specialize the code anyways" idea can be applied to most code, but I would suggest that the wins in readability and maintainability are more important over time.

Basically, I would suggest resisting the urge to ever Ctrl+C/V code and instead go for a refactoring where a more general concept is "discovered". I personally tend to refactor at reasonably sized and well-defined boundaries even before reaching the 2nd instance of something.

Reply:
Makes sense :)